# Secret of Best Hacker News Post

- In this project the data provided by Hacker News will be analyzed. You can find the dataset [here](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts).
- More specifically two types of post will be compared to find which receives more comments on average. ``Ask HN`` or ``Show HN``
- Lastly, how does time factor affects the number comments on average

In [2]:
from csv import reader

In [3]:
def extract_data(file):
    """locates file using filepath, reads it, then returns a list of lists"""
    
    opened_file = open(file)
    read_file = reader(opened_file)
    result = list(read_file)
    
    # return both the column headers and the actual data as seperate entities
    return result[0], result[1:]

In [4]:
header, hn = extract_data('hacker_news.csv')

In [6]:
header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [7]:
def explore_data(dataset, start, end, n_rows_columns = False):
    
    dataset_slice = dataset[start:end]
    
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if n_rows_columns:
        print('The number of rows: {:,}'.format(len(dataset)))
        print('The number of columns: ', len(dataset[0]))
              
              
#           if rows_and_columns:
#         print('Number of rows: {:,}'.format(len(dataset)))
#         print('Number of columns:', len(dataset[0]))

In [19]:
explore_data(hn, 0, 5, n_rows_columns = True)

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']


The number of rows: 20,100
The number of columns:  7


In [8]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)


In [9]:
template = 'The number of {} posts are: {:,} ({:.0f}%)'
    
    
print(template.format('ask hn', len(ask_posts), len(ask_posts)/len(hn)* 100))
print(template.format('show hn', len(show_posts), len(show_posts)/len(hn)* 100))
print(template.format('other', len(other_posts), len(other_posts)/len(hn)* 100))


The number of ask hn posts are: 1,744 (9%)
The number of show hn posts are: 1,162 (6%)
The number of other posts are: 17,194 (86%)


In [10]:
total_ask_comments = 0
for post in ask_posts:
    n_comments = int(post[4])
    total_ask_comments += n_comments
avg_ask_comments = total_ask_comments / len(ask_posts)
print("The average num of comment in ask posts is: {:.2f}".format(avg_ask_comments))

total_show_comments = 0
for post in show_posts:
    n_comments = int(post[4])
    total_show_comments += n_comments
avg_show_comments = total_show_comments / len(show_posts)
print("The average number of comments in show posts is: {:.2f}".format(avg_show_comments))


The average num of comment in ask posts is: 14.04
The average number of comments in show posts is: 10.32


- As we see. ``Ask posts`` receive on average more comments than their ``Show`` counterparts. **14.04** VS **10.32**

In [11]:
import datetime as dt

In [37]:
result_list = []
for post in ask_posts:
    result_list_lite = []
    create_at = post[6]
    n_comments = int(post[4])

    result_list.append([create_at, n_comments])
    
    
counts_by_hour = {}
comments_by_hour = {}
    
for row in result_list:
    comments = int(row[1])
    date = row[0]
    date = dt.datetime.strptime(date, '%m/%d/%Y %H:%M')
    hour = date.hour
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments

for key in counts_by_hour:
    print(key, ": ", counts_by_hour[key])
    
print("\n")
for key in comments_by_hour:
    print(key, ": ", comments_by_hour[key])
    
    
    import datetime as dt

result_list = []

for post in ask_posts:
    comments = int(post[4])
    created = post[6]
    result_list.append([created, comments])

#########################################################################    
    
# counts_by_hour = {}
# comments_by_hour = {}

# for item in result_list:
#     created = dt.datetime.strptime(item[0], '%m/%d/%Y %H:%M')
#     hour_created = created.hour
#     comments = int(item[1])
#     if hour_created in counts_by_hour:
#         counts_by_hour[hour_created] += 1
#         comments_by_hour[hour_created] += comments
#     else:
#         counts_by_hour[hour_created] = 1
#         comments_by_hour[hour_created] = comments

# print('Posts count per hour:', counts_by_hour)
# print('Comments count per hour:', comments_by_hour)


# avg_by_hour = []

# for hour in comments_by_hour:
#     avg_by_hour.append([hour, round(comments_by_hour[hour] / counts_by_hour[hour], 1)])
    
# print('Average number of comments by hour:', avg_by_hour)



9 :  45
13 :  85
10 :  59
14 :  107
16 :  108
23 :  68
12 :  73
17 :  100
15 :  116
21 :  109
20 :  80
2 :  58
18 :  109
3 :  54
5 :  46
19 :  110
1 :  60
22 :  71
8 :  48
4 :  47
0 :  55
6 :  44
7 :  34
11 :  58


9 :  251
13 :  1253
10 :  793
14 :  1416
16 :  1814
23 :  543
12 :  687
17 :  1146
15 :  4477
21 :  1745
20 :  1722
2 :  1381
18 :  1439
3 :  421
5 :  464
19 :  1188
1 :  683
22 :  479
8 :  492
4 :  337
0 :  447
6 :  397
7 :  267
11 :  641


In [38]:

avg_by_hour = []
for key in counts_by_hour:
    avg_count = int(comments_by_hour[key]) / int(counts_by_hour[key]) 
    avg_by_hour.append([key, avg_count])

print(avg_by_hour)

[[9, 5.5777777777777775], [13, 14.741176470588234], [10, 13.440677966101696], [14, 13.233644859813085], [16, 16.796296296296298], [23, 7.985294117647059], [12, 9.41095890410959], [17, 11.46], [15, 38.5948275862069], [21, 16.009174311926607], [20, 21.525], [2, 23.810344827586206], [18, 13.20183486238532], [3, 7.796296296296297], [5, 10.08695652173913], [19, 10.8], [1, 11.383333333333333], [22, 6.746478873239437], [8, 10.25], [4, 7.170212765957447], [0, 8.127272727272727], [6, 9.022727272727273], [7, 7.852941176470588], [11, 11.051724137931034]]


In [39]:
swap_avg_by_hour = []
for row in avg_by_hour:
    hour, count = row[0], row[1]
    swap_avg_by_hour.append([count, hour])
print(swap_avg_by_hour)

[[5.5777777777777775, 9], [14.741176470588234, 13], [13.440677966101696, 10], [13.233644859813085, 14], [16.796296296296298, 16], [7.985294117647059, 23], [9.41095890410959, 12], [11.46, 17], [38.5948275862069, 15], [16.009174311926607, 21], [21.525, 20], [23.810344827586206, 2], [13.20183486238532, 18], [7.796296296296297, 3], [10.08695652173913, 5], [10.8, 19], [11.383333333333333, 1], [6.746478873239437, 22], [10.25, 8], [7.170212765957447, 4], [8.127272727272727, 0], [9.022727272727273, 6], [7.852941176470588, 7], [11.051724137931034, 11]]


In [41]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print(sorted_swap)

[[38.5948275862069, 15], [23.810344827586206, 2], [21.525, 20], [16.796296296296298, 16], [16.009174311926607, 21], [14.741176470588234, 13], [13.440677966101696, 10], [13.233644859813085, 14], [13.20183486238532, 18], [11.46, 17], [11.383333333333333, 1], [11.051724137931034, 11], [10.8, 19], [10.25, 8], [10.08695652173913, 5], [9.41095890410959, 12], [9.022727272727273, 6], [8.127272727272727, 0], [7.985294117647059, 23], [7.852941176470588, 7], [7.796296296296297, 3], [7.170212765957447, 4], [6.746478873239437, 22], [5.5777777777777775, 9]]


In [49]:
for row in sorted_swap[:5]:
    hour = dt.datetime.strptime(str(row[1]), '%H')
    hour = hour.strftime('%H:%M')
    print("{}: {:.2f} average comments per post".format(hour, row[0]))

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
